### ISSUE TITLE
investigate restarting of networks and traffic incurred 

### ISSUE THREAD

**Issue by acud** [ethersphere/swarm#1396]
when we restart a swarm node the kademlia table's connection changes - nodes which are known about and that were with established connections to us are not prioritized over other nodes.

this, in turn, incurs a big traffic load as a node suddenly has to push for example half of its chunks (bin 0) to all of the new peers in bin 0

this is easily reproducible when running a persistent cluster, shutting it down, then starting the nodes again.

we should prioritize known peers when reconnecting

**Comment by nonsense** [comment]
when we start a swarm node, we should know what our previous kademlia table was (who our peers were) and peers that we were connected to should have priority over peers we `know` about.

**Comment by nonsense** [comment]
~when we look into this we should also consider how to reduce the number of redundant peers - sometimes we have a lot of peers in bin 0, 1, etc. - we should have only K peers (where K is 2 or 3 or other small value), rather than 10 or 20 peers in lower bins.~

peers exchange information regarding their view of the network (the hive protocol), if they also say how many peers they are connected to in a given bin, we could maybe start dropping redundant connections/peers.

**Comment by kortatu** [comment]
> when we look into this we should also consider how to reduce the number of redundant peers - sometimes we have a lot of peers in bin 0, 1, etc. - we should have only K peers (where K is 2 or 3 or other small value), rather than 10 or 20 peers in lower bins.
> 

How this relates  to #1436 ?
In PR #1833, we have implemented a higher number of peers for lowe bins (0,1, etc..) than for higher bins.

**Comment by nonsense** [comment]
@kortatu good question, sorry about the confusion.

Currently by design, we should have only K peers in every bucket. This is what is mentioned in this issue here.

However as explained in #1436 , this is not a great solution because:

```
Make sure we have a lot of peers in bin 0, as they are responsible for retrieve requests for half of the chunks on any given download, and we should run fetch requests to all of them. For example, if we are supposed to fetch a 4GB file in a timely fashion, then 1/2 * 4GB == 2GB chunks will be retrieved from bin 0 peers, then we need at least 10 peers for bin 0, in order to deliver adequate user experience.
```

I believe the correct solution is to use `more` peers in lower bins and load balance on them (according to XOR, or round-robin, or whatever).

I will check #1833 but based on what you say, you've implemented a better strategy than the one we currently have and discussed here.

**Comment by nonsense** [comment]
@kortatu then again, ideally we should have functionality to drop redundant peers and some way of determining which peer connection is redundant.

**Comment by nonsense** [comment]
@kortatu you are right about that. There is not guarantee that the space will be split between the peers in a given bin, and they can be skewed a lot.

**Comment by kortatu** [comment]
And if the case we have the three peers `10000000`, `10000001` and `11000000`, one of the first two will be redundant as you said.

And other consequence of the address space is that, by looking at the stats of connections per peers, we can decide to get a new peer near the the ones more overloaded. So if the previous example, I have much more connections to `11000000` we should consider looking for a peer near that one , ideally `11100000` as it will cover half the space previously covered by `11000000` (so if we find some overloaded peer with a PO of N, we should add peer with a proximity order of N+1, N+2... to it )
